overlord/patch/patch.go

```diff
@@ -40,45 +49,147 @@ func Init(s *state.State) {
 		panic("internal error: expected empty state, attempting to override patch-level without actual patching")
 	}
 	s.Set("patch-level", Level)
+
+	if s.Get("patch-sublevel", new(int)) != state.ErrNoState {
+		panic("internal error: expected empty state, attempting to override patch-sublevel without actual patching")
+	}
+	s.Set("patch-sublevel", Sublevel)
+}
+
+// applySublevelPatches applies all sublevel patches for given level, starting
+// from firstSublevel index.
+func applySublevelPatches(level, firstSublevel int, s *state.State) error {
+	for sublevel := firstSublevel; sublevel < len(patches[level]); sublevel++ {
+		if sublevel > 0 {
+			logger.Noticef("Patching system state level %d to sublevel %d...", level, sublevel)
+		}
+		err := applyOne(patches[level][sublevel], s, level, sublevel)
+		if err != nil {
+			logger.Noticef("Cannot patch: %v", err)
+			return fmt.Errorf("cannot patch system state to level %d, sublevel %d: %v", level, sublevel, err)
+		}
+	}
+	return nil
+}
+
+// maybeResetSublevelForLevel60 checks if the previously installed core revision was a patch 6 sublevel 0
+// i.e. unaware of sublevels and the need for resetting sublevel state and if so, resets sublevel to 0
+// to re-apply sublevel patches.
+func maybeResetSublevelForLevel60(s *state.State, sublevel *int) error {
+	s.Lock()
+	defer s.Unlock()
+
+	core, err := snapstate.CoreInfo(s)
+	if err == state.ErrNoState {
+		// no core snap - nothing to do.
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	var snapst snapstate.SnapState
+	if err := snapstate.Get(s, core.InstanceName(), &snapst); err != nil {
+		return err
+	}
+	if len(snapst.Sequence) < 2 {
+		return nil
+	}
+	currentIndex := snapst.LastIndex(snapst.Current)
+	if currentIndex < 0 {
+		return fmt.Errorf("internal error: couldn't find current core revision in the snap sequence")
+	}
+
+	prevRev := snapst.Sequence[currentIndex-1].Revision
+
+	// check if core revision is for patch level 6.0.
+	// the revision number here is the highest rev number from candidate channels of all architectures
+	if prevRev.N > 5332 {
+		return nil
+	}
+
+	var sublevelResetTime time.Time
+	var lastRefresh time.Time
+	if err := s.Get("last-refresh", &lastRefresh); err != nil && err != state.ErrNoState {
```

Comment:
last-refresh is the last auto-refresh only, is not the last core refresh time, the closest to that would be the /snap/core/current symlink time